File: backend/src/services/googlePlacesService.js
import axios from 'axios';

const GOOGLE_MAPS_API_KEY = process.env.GOOGLE_MAPS_API_KEY;
const PLACES_API_URL = 'https://maps.googleapis.com/maps/api/place';

/**
 * Business type mapping to Google Places types
 * Reference: https://developers.google.com/maps/documentation/places/web-service/supported_types
 */
const BUSINESS_TYPE_MAPPING = {
  grocery: ['grocery_or_supermarket', 'supermarket'],
  cafe: ['cafe', 'coffee_shop'],
  hotel: ['lodging'],  // lodging covers hotels, motels, etc.
  dairy: ['grocery_or_supermarket', 'store'],
  restaurant: ['restaurant', 'meal_takeaway', 'meal_delivery'],
  bakery: ['bakery'],
  pharmacy: ['pharmacy'],
  supermarket: ['supermarket', 'grocery_or_supermarket'],
  hostel: ['lodging'],  // hostel is also lodging type
  schools: ['school', 'primary_school', 'secondary_school'],
  colleges: ['university'],
  hospitals: ['hospital', 'doctor'],
  others: ['store', 'establishment']
};

/**
 * Get coordinates for a pincode using Geocoding API
 */
export const getCoordinatesFromPincode = async (pincode, country = 'India') => {
  try {
    const response = await axios.get(`${PLACES_API_URL}/textsearch/json`, {
      params: {
        query: `${pincode}, ${country}`,
        key: GOOGLE_MAPS_API_KEY
      }
    });

    if (response.data.results && response.data.results.length > 0) {
      const location = response.data.results[0].geometry.location;
      return {
        success: true,
        latitude: location.lat,
        longitude: location.lng
      };
    }

    return { success: false, message: 'Location not found' };
  } catch (error) {
    console.error('Geocoding error:', error.message);
    return { success: false, message: 'Failed to get coordinates', error: error.message };
  }
};

/**
 * Search for businesses near a location
 */
export const searchBusinessesNearby = async (latitude, longitude, businessType, radius = 5000) => {
  try {
    // Validate API key
    if (!GOOGLE_MAPS_API_KEY) {
      return {
        success: false,
        message: 'Google Maps API key not configured',
        businesses: []
      };
    }

    const types = BUSINESS_TYPE_MAPPING[businessType.toLowerCase()] || ['store'];
    const allResults = [];

    console.log(`🔍 Searching for ${businessType} near (${latitude}, ${longitude})`);
    console.log(`📋 Using Google Places types: ${types.join(', ')}`);

    for (const type of types) {
      try {
        const response = await axios.get(`${PLACES_API_URL}/nearbysearch/json`, {
          params: {
            location: `${latitude},${longitude}`,
            radius: radius,
            type: type,
            key: GOOGLE_MAPS_API_KEY
          }
        });

        if (response.data.status === 'OK' && response.data.results) {
          console.log(`✅ Found ${response.data.results.length} results for type: ${type}`);
          allResults.push(...response.data.results);
        } else if (response.data.status === 'ZERO_RESULTS') {
          console.log(`⚠️ No results for type: ${type}`);
        } else {
          console.log(`⚠️ API error for type ${type}: ${response.data.status} - ${response.data.error_message || ''}`);
        }
      } catch (error) {
        console.error(`❌ Error searching type ${type}:`, error.message);
      }

      // Add delay to avoid rate limiting
      await new Promise(resolve => setTimeout(resolve, 200));
    }

    // Remove duplicates based on place_id
    const uniqueResults = Array.from(
      new Map(allResults.map(item => [item.place_id, item])).values()
    );

    console.log(`📊 Total unique results: ${uniqueResults.length}`);

    return {
      success: true,
      businesses: uniqueResults.map(place => ({
        placeId: place.place_id,
        name: place.name,
        address: place.vicinity,
        latitude: place.geometry?.location?.lat,
        longitude: place.geometry?.location?.lng,
        rating: place.rating,
        userRatingsTotal: place.user_ratings_total,
        businessType: businessType,
        openNow: place.opening_hours?.open_now,
        photos: place.photos?.map(p => ({ photoReference: p.photo_reference })) || [],
        priceLevel: place.price_level
      }))
    };
  } catch (error) {
    console.error('Places search error:', error.message);
    return {
      success: false,
      message: 'Failed to search businesses',
      error: error.message,
      businesses: []
    };
  }
};

/**
 * Get place details by place ID
 */
export const getPlaceDetails = async (placeId) => {
  try {
    const response = await axios.get(`${PLACES_API_URL}/details/json`, {
      params: {
        place_id: placeId,
        fields: 'name,formatted_address,formatted_phone_number,geometry,rating,types,reviews,photos,opening_hours,user_ratings_total,price_level,website',
        key: GOOGLE_MAPS_API_KEY
      }
    });

    if (response.data.result) {
      const place = response.data.result;
      
      // Format photos with photo references
      const photos = place.photos?.map(photo => ({
        photoReference: photo.photo_reference,
        width: photo.width,
        height: photo.height
      })) || [];

      // Format reviews
      const reviews = place.reviews?.map(review => ({
        authorName: review.author_name,
        profilePhotoUrl: review.profile_photo_url,
        rating: review.rating,
        text: review.text,
        relativeTimeDescription: review.relative_time_description,
        time: review.time
      })) || [];

      return {
        success: true,
        place: {
          ...place,
          photos,
          reviews,
          userRatingsTotal: place.user_ratings_total,
          priceLevel: place.price_level,
          openNow: place.opening_hours?.open_now,
          openingHours: place.opening_hours?.weekday_text
        }
      };
    }

    return { success: false, message: 'Place not found' };
  } catch (error) {
    console.error('Place details error:', error.message);
    return { success: false, message: 'Failed to get place details', error: error.message };
  }
};

/**
 * Search businesses by pincode and business types
 */
export const searchBusinessesByPincode = async (pincode, businessTypes, areas = []) => {
  try {
    console.log(`\n🔍 Searching businesses in pincode: ${pincode}`);
    console.log(`📋 Business types requested: ${businessTypes.join(', ')}`);

    // Validate API key
    if (!GOOGLE_MAPS_API_KEY) {
      console.error('❌ Google Maps API key not configured');
      return {
        success: false,
        message: 'Google Maps API key not configured',
        businesses: []
      };
    }

    // Get coordinates for pincode
    const coordsResult = await getCoordinatesFromPincode(pincode);
    if (!coordsResult.success) {
      console.log(`❌ Could not geocode pincode ${pincode}: ${coordsResult.message}`);
      return {
        success: false,
        message: `Could not find location for pincode ${pincode}`,
        businesses: []
      };
    }

    const { latitude, longitude } = coordsResult;
    console.log(`📍 Coordinates: ${latitude}, ${longitude}`);

    const allBusinesses = [];
    const breakdown = {};

    // Search for each business type (convert to lowercase for mapping)
    for (const businessType of businessTypes) {
      const normalizedType = businessType.toLowerCase();
      console.log(`\n🔎 Searching for: ${businessType} (normalized: ${normalizedType})`);

      try {
        const result = await searchBusinessesNearby(latitude, longitude, normalizedType);
        if (result.success && result.businesses) {
          allBusinesses.push(...result.businesses);
          breakdown[businessType] = result.businesses.length;
          console.log(`✅ Found ${result.businesses.length} ${businessType} businesses`);
        } else {
          console.log(`⚠️ Failed to search ${businessType}: ${result.message || 'Unknown error'}`);
          breakdown[businessType] = 0;
        }
      } catch (error) {
        console.error(`❌ Error searching ${businessType}:`, error.message);
        breakdown[businessType] = 0;
      }
    }

    // Remove duplicates based on placeId
    const uniqueBusinesses = [];
    const seenPlaceIds = new Set();

    for (const business of allBusinesses) {
      if (business.placeId && !seenPlaceIds.has(business.placeId)) {
        seenPlaceIds.add(business.placeId);
        uniqueBusinesses.push(business);
      }
    }

    console.log(`\n📊 Total businesses found: ${uniqueBusinesses.length}`);
    console.log(`📊 Breakdown:`, breakdown);

    return {
      success: true,
      totalBusinesses: uniqueBusinesses.length,
      breakdown: breakdown,
      businesses: uniqueBusinesses,
      message: `Found ${uniqueBusinesses.length} businesses in pincode ${pincode}`
    };
  } catch (error) {
    console.error('❌ Business search error:', error.message);
    return {
      success: false,
      message: 'Failed to search businesses',
      error: error.message,
      businesses: []
    };
  }
};

File: loagma_crm/lib/services/google_places_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/google_places_config.dart';

class GooglePlacesService {
  static const String _baseUrl = 'https://maps.googleapis.com/maps/api/place';

  /// Get place details including reviews and photos
  static Future<Map<String, dynamic>?> getPlaceDetails(String placeId) async {
    if (!GooglePlacesConfig.isConfigured) {
      print('⚠️ Google Places API key not configured');
      print('📝 To enable Google Places features:');
      print(
        '   1. Get an API key from https://console.cloud.google.com/apis/credentials',
      );
      print('   2. Enable the Places API for your project');
      print('   3. Set your API key in GooglePlacesConfig.apiKey');
      return null;
    }

    try {
      final url = Uri.parse(
        '$_baseUrl/details/json?place_id=$placeId&fields=name,rating,reviews,photos,formatted_phone_number,opening_hours,website,price_level&key=${GooglePlacesConfig.apiKey}',
      );

      final response = await http.get(url);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        if (data['status'] == 'OK') {
          return data['result'];
        } else {
          print('Google Places API error: ${data['status']}');
          return null;
        }
      } else {
        print('HTTP error: ${response.statusCode}');
        return null;
      }
    } catch (e) {
      print('Error fetching place details: $e');
      return null;
    }
  }

  /// Fetch nearby places using Google Places API
  static Future<List<Map<String, dynamic>>> fetchNearbyPlaces({
    required double lat,
    required double lng,
    required int radius,
    required String type,
  }) async {
    if (!GooglePlacesConfig.isConfigured) {
      print('⚠️ Google Places API key not configured for nearby places');
      return [];
    }

    try {
      final url = Uri.parse(
        '$_baseUrl/nearbysearch/json?location=$lat,$lng&radius=$radius&type=$type&key=${GooglePlacesConfig.apiKey}',
      );

      final response = await http.get(url);

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        if (data['status'] == 'OK') {
          return List<Map<String, dynamic>>.from(data['results'] ?? []);
        } else {
          print('Google Places API error: ${data['status']}');
          return [];
        }
      } else {
        print('HTTP error: ${response.statusCode}');
        return [];
      }
    } catch (e) {
      print('Error fetching nearby places: $e');
      return [];
    }
  }

  /// Fetch place details (alias for getPlaceDetails for compatibility)
  static Future<Map<String, dynamic>?> fetchPlaceDetails(String placeId) async {
    return getPlaceDetails(placeId);
  }

  /// Get photo URL from photo reference
  static String getPhotoUrl(String photoReference, {int maxWidth = 400}) {
    if (!GooglePlacesConfig.isConfigured) {
      return '';
    }
    return '$_baseUrl/photo?maxwidth=$maxWidth&photo_reference=$photoReference&key=${GooglePlacesConfig.apiKey}';
  }

  /// Format reviews for display
  static List<Map<String, dynamic>> formatReviews(List<dynamic>? reviews) {
    if (reviews == null) return [];

    return reviews.map((review) {
      return {
        'author_name': review['author_name'] ?? 'Anonymous',
        'rating': review['rating'] ?? 0,
        'text': review['text'] ?? '',
        'time': review['time'] ?? 0,
        'profile_photo_url': review['profile_photo_url'],
        'relative_time_description': review['relative_time_description'] ?? '',
      };
    }).toList();
  }

  /// Format photos for display
  static List<String> formatPhotos(
    List<dynamic>? photos, {
    int maxWidth = 400,
  }) {
    if (photos == null) return [];

    return photos.map((photo) {
      final photoReference = photo['photo_reference'];
      return getPhotoUrl(photoReference, maxWidth: maxWidth);
    }).toList();
  }

  /// Get opening hours text
  static List<String> getOpeningHours(Map<String, dynamic>? openingHours) {
    if (openingHours == null || openingHours['weekday_text'] == null) {
      return [];
    }
    return List<String>.from(openingHours['weekday_text']);
  }

  /// Format price level
  static String formatPriceLevel(int? priceLevel) {
    if (priceLevel == null) return 'Price not available';

    switch (priceLevel) {
      case 0:
        return 'Free';
      case 1:
        return 'Inexpensive (\$)';
      case 2:
        return 'Moderate (\$\$)';
      case 3:
        return 'Expensive (\$\$\$)';
      case 4:
        return 'Very Expensive (\$\$\$\$)';
      default:
        return 'Price not available';
    }
  }
}
